### PR TITLE
Add Hadoop compatible OutputStream

### PR DESCRIPTION
**Pull request by winston-zillow**
Even though hadoop uses snappy-java, its snappy files are not compatible with the ones produced by snapp-yjava's OutputStream due to header and block preemble differences. This PR adds a hadoop-compatible output stream so that we can create snappy files outside of hadoop and yet still be able to consumed by hadoop.

### CODE REVIEW

File: src/main/java/org/xerial/snappy/SnappyOutputStream.java [hunk 1]
```diff
@@ -102,7 +108,9 @@ public SnappyOutputStream(OutputStream out, int blockSize, BufferAllocatorFactor
         inputBuffer = inputBufferAllocator.allocate(inputSize);
         outputBuffer = outputBufferAllocator.allocate(outputSize);
 
-        outputCursor = SnappyCodec.currentHeader.writeHeader(outputBuffer, 0);
+        if (writeHeader) {
```
Comment (xerial) [inline src/main/java/org/xerial/snappy/SnappyOutputStream.java]: Can we implement this by function overload?  For example, defining a protected method (writeHeader()), then override this in Hadoop stream side.

File: src/main/java/org/xerial/snappy/SnappyHadoopCompatibleOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package org.xerial.snappy;
+
+import java.io.OutputStream;
+
+import org.xerial.snappy.buffer.BufferAllocatorFactory;
+import org.xerial.snappy.buffer.CachedBufferAllocator;
+
+public class SnappyHadoopCompatibleOutputStream extends SnappyOutputStream
+{
+	public SnappyHadoopCompatibleOutputStream(OutputStream out)
```
Comment (xerial) [inline src/main/java/org/xerial/snappy/SnappyHadoopCompatibleOutputStream.java]: 4-space indentation is preferred

File: src/main/java/org/xerial/snappy/SnappyOutputStream.java [hunk 3]
```diff
@@ -91,6 +91,12 @@ public SnappyOutputStream(OutputStream out, int blockSize)
 
     public SnappyOutputStream(OutputStream out, int blockSize, BufferAllocatorFactory bufferAllocatorFactory)
     {
+        this(out, blockSize, bufferAllocatorFactory, true);
+    }
+
+    protected SnappyOutputStream(OutputStream out, int blockSize, BufferAllocatorFactory bufferAllocatorFactory,
+            boolean writeHeader)
```
Comment (xerial) [inline src/main/java/org/xerial/snappy/SnappyOutputStream.java]: adding boolean argument is not recommended.

**Comment by xerial** [comment]
Interesting. 

- Indent should be 4-white spaces
- Need to find a way to pass test cases
- Adding boolean flag to the SnappyOutputStream constructor might not be a good way. Overriding method seems much cleaner.

**Comment by winston-zillow** [comment]
The missing testing is hard to fix. Do you accept if it just does try-catch the link error and skip the hadoop read-back test?

**Comment by winston-zillow** [comment]
I can get it to pass on my mac OS X and Ubuntu 16.04

```
[info] Test run finished: 0 failed, 0 ignored, 17 total, 0.459s
[info] Test run started
copying /lib/Linux/libhadoop.so => /tmp/xerial-snappy-test3485064251324217366.nativelib/libhadoop.so
copying /lib/Linux/libsnappy.so => /tmp/xerial-snappy-test3485064251324217366.nativelib/libsnappy.so
copying /lib/Linux/libsnappy.so => /tmp/xerial-snappy-test3485064251324217366.nativelib/libsnappy.so.1
LD_LIBRARY_PATH = /tmp/xerial-snappy-test3485064251324217366.nativelib:null
   copied /tmp/xerial-snappy-test3485064251324217366.nativelib/libhadoop.so size: 800056
   copied /tmp/xerial-snappy-test3485064251324217366.nativelib/libsnappy.so.1 size: 30904
   copied /tmp/xerial-snappy-test3485064251324217366.nativelib/libsnappy.so size: 30904
[info] Test org.xerial.snappy.SnappyHadoopCompatibleOutputStreamTest.testXerialCompressionHadoopDecompressionCodec started
[info] Test run finished: 0 failed, 0 ignored, 1 total, 0.458s
[info] Test run started
```
OS and library info
```
root@21608b7b98a9:~/snappy-java# uname -a 
Linux 21608b7b98a9 4.9.27-moby #1 SMP Thu May 11 04:01:18 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
root@21608b7b98a9:~/snappy-java# cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
root@21608b7b98a9:~/snappy-java# strings /usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
GLIBCXX_3.4.3
GLIBCXX_3.4.4
GLIBCXX_3.4.5
GLIBCXX_3.4.6
GLIBCXX_3.4.7
GLIBCXX_3.4.8
GLIBCXX_3.4.9
GLIBCXX_3.4.10
GLIBCXX_3.4.11
GLIBCXX_3.4.12
GLIBCXX_3.4.13
GLIBCXX_3.4.14
GLIBCXX_3.4.15
GLIBCXX_3.4.16
GLIBCXX_3.4.17
GLIBCXX_3.4.18
GLIBCXX_3.4.19
GLIBCXX_3.4.20
GLIBCXX_3.4.21
GLIBCXX_DEBUG_MESSAGE_LENGTH
root@21608b7b98a9:~/snappy-java# ldd src/test/resources/lib/Linux/libhadoop.so 
	linux-vdso.so.1 =>  (0x00007fffbcd80000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc8eb2cd000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc8eaf04000)
	/lib64/ld-linux-x86-64.so.2 (0x000055b0724a7000)
root@21608b7b98a9:~/snappy-java# ldd src/test/resources/lib/Linux/libsnappy.so 
	linux-vdso.so.1 =>  (0x00007ffdd479d000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f686def1000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f686db28000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f686d911000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f686d608000)
	/lib64/ld-linux-x86-64.so.2 (0x00005579642b5000)
```
What do you use in this Travis CI build?

**Comment by winston-zillow** [comment]
@xerial any response?

**Comment by winston-zillow** [comment]
@xerial I eventually won the battle against the Linux shared library! Please take a look.

File: src/test/java/org/xerial/snappy/SnappyHadoopCompatibleOutputStreamTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            libraryNames.put("libhadoop.so", "libhadoop.so");
+            // certain Linux systems need these shared library be copied before the JVM started, see build.sbt
+            libraryNames.put("libsnappy.so", "libsnappy.so");
+            libraryNames.put("libsnappy.so.1", "libsnappy.so");
+        } else if (SystemUtils.IS_OS_MAC_OSX) {
+            libResourceFolder = "/lib/MacOSX";
+            libraryNames.put("libhadoop.dylib", "libhadoop.dylib");
+            libraryNames.put("libsnappy.dylib", "libsnappy.dylib");
+            libraryNames.put("libsnappy.1.dylib", "libsnappy.dylib");
+        } else {
```
Comment (winston-zillow) [inline src/test/java/org/xerial/snappy/SnappyHadoopCompatibleOutputStreamTest.java]: Note: we could do all of these in the sbt, but I feel we should keep the sbt simple. Eventually when the Linux systems have all been upgraded, we could do away those copying in sbt.

**Comment by xerial** [comment]
@winston-zillow OK. I'll have a look. Thanks

**Comment by winston-zillow** [comment]
hi, will this one be released soon? thanks

**Comment by xerial** [comment]
Merged in #186. @winston-zillow Sorry for keeping you waiting for a long time.

**Pull request closed without merging**